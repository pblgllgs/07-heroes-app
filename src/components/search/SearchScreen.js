import React from 'react'
import { useForm } from '../../hooks/useForm-';

export const SearchScreen = () => {

    const initialState= {
        searchText: ''
    }
    const [{searchText}, handleInputChange, reset] = useForm(initialState);

    const handleSearch = (e) => {
        e.preventDefault()
        console.log(searchText)
        reset();
    }

    return (
        <>
            <h1>Busqueda</h1>
            <hr />
            <div className='row'>
                <div className='col-5'>
                    <h4>Buscar</h4>
                    <hr />
                    <form onSubmit={handleSearch}>
                        <input
                            type='text'
                            placeholder='Ingresa un id'
                            className='form-control'
                            name='searchText'
                            autoComplete='off'
                            value={searchText}
                            onChange={handleInputChange}
                        />
                        <button className='btn btn-primary mt-3' type='submit'>
                            Buscar
                        </button>
                    </form>
                </div>
            </div>
        </>
    )
}
