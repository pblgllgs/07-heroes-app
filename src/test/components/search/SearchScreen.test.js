import { SearchScreen } from "../../../components/search/SearchScreen";
import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";

//creacion de un mock para simular el uso de navigate
const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => mockNavigate,
}));

describe("Pruebas en <SearchScreen />", () => {
    test("debe mostrarse correctamente con valores por defecto", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/"]}>
                <SearchScreen />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".alert-info").text().trim()).toBe(
            "Acá se mostrarán las coincidencias"
        );
    });

    test("debe mostrarse correctamente a batman y el input del queryString", () => {
        //capturamos el componente montado
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <SearchScreen />
            </MemoryRouter>
        );
        //se compara con el snapshot
        expect(wrapper).toMatchSnapshot();
        //se espera que el valor del item input sea batman
        expect(wrapper.find("input").prop("value")).toBe("batman");
        //se espera que el valor del item dentro de card-title sea Batman
        expect(wrapper.find(".card-title").text().trim()).toBe("Batman");
    });

    test("debe mostrarse el error con el valor batman123 y el input del queryString debe ser 123", () => {
        //capturamos el componente montado
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman123"]}>
                <SearchScreen />
            </MemoryRouter>
        );
        //se compara con el snapshot
        expect(wrapper).toMatchSnapshot();
        //se espera que el valor del item input sea batman
        expect(wrapper.find("input").prop("value")).toBe("batman123");
        //se espera que el valor del item dentro de card-title sea Batman
        expect(wrapper.find(".alert-danger").text().trim()).toBe(
            "No hay resultados para batman123"
        );
    });

    test("debe de llamar el navigate a la nueva pantalla", () => {
        //capturamos el componente montado
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <SearchScreen />
            </MemoryRouter>
        );
        //se simula el cambio del formulario con el valor de batman
        wrapper.find("input").simulate("change", {
            target: {
                name: "searchText",
                value: "batman",
            },
        });
        //se simula el posteo del form
        wrapper.find("form").prop("onSubmit")({
            preventDefault() {},
        });
        //se espera que el navigate se haga con el valor enviado en el target
        expect(mockNavigate).toHaveBeenCalledWith("?q=batman");
    });
});
