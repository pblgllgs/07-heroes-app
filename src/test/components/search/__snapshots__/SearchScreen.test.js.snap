// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe mostrarse correctamente a batman y el input del queryString 1`] = `
Array [
  <h1>
    Busqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Ingresa un nombre..."
          type="text"
          value="batman"
        />
        <button
          className="btn btn-primary mt-3"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchScreen /> debe mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Busqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Ingresa un nombre..."
          type="text"
          value=""
        />
        <button
          className="btn btn-primary mt-3"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Acá se mostrarán las coincidencias
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchScreen /> debe mostrarse el error con el valor batman123 y el input del queryString debe ser 123 1`] = `
Array [
  <h1>
    Busqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Ingresa un nombre..."
          type="text"
          value="batman123"
        />
        <button
          className="btn btn-primary mt-3"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        No hay resultados para 
        batman123
      </div>
    </div>
  </div>,
]
`;
